eZ/Publish/Core/Repository/UserService.php

```diff
@@ -770,6 +748,108 @@ public function updateUser(APIUser $user, UserUpdateStruct $userUpdateStruct): A
+        ) {
+            throw new UnauthorizedException('user', 'password');
+        }
+
+        $userFieldDefinition = $this->getUserFieldDefinition($loadedUser->getContentType());
+        if ($userFieldDefinition === null) {
+            throw new MissingUserFieldTypeException($loadedUser->getContentType());
+        }
+
+        $contentUpdateStruct = $contentService->newContentUpdateStruct();
+        $contentUpdateStruct->setField(
+            $userFieldDefinition->identifier,
+            new UserValue([
+                'contentId' => $loadedUser->id,
+                'hasStoredLogin' => true,
+                'login' => $loadedUser->login,
+                'email' => $loadedUser->email,
+                'plainPassword' => $newPassword,
+                'enabled' => $loadedUser->enabled,
+                'maxLogin' => $loadedUser->maxLogin,
+                'passwordHashType' => $user->hashAlgorithm,
+                'passwordHash' => $user->passwordHash,
+            ])
+        );
+
+        $errors = $this->contentValidator->validate(
+            $contentUpdateStruct,
+            ['content' => $loadedUser],
+        );
+
+        if (!empty($errors)) {
+            throw new ContentFieldValidationException($errors);
+        }
+
+        $contentType = $this->repository->getContentTypeService()->loadContentType(
+            $loadedUser->contentInfo->contentTypeId
+        );
+
+        $fields = $this->contentMapper->mapFieldsForUpdate(
+            $contentUpdateStruct,
+            $contentType,
+            $userFieldDefinition->mainLanguageCode
+        );
+
+        $fieldValue = $this->contentMapper->getFieldValueForUpdate(
+            $fields[$userFieldDefinition->identifier][$userFieldDefinition->mainLanguageCode],
+            $loadedUser->getField($userFieldDefinition->identifier, $userFieldDefinition->mainLanguageCode),
+            $userFieldDefinition,
+            false
+        );
+
+        $fieldType = $this->fieldTypeRegistry->getFieldType(
+            $userFieldDefinition->fieldTypeIdentifier
+        );
+        $value = $fieldType->toPersistenceValue($fieldValue);
+
+        $this->repository->beginTransaction();
+        try {
+            $this->userHandler->updatePassword(
+                new SPIUser(
+                    [
+                        'id' => $loadedUser->id,
+                        'login' => $loadedUser->login,
+                        'email' => $loadedUser->email,
+                        'passwordHash' => $value->externalData['passwordHash'],
+                        'hashAlgorithm' => $value->externalData['passwordHashType'],
+                        'passwordUpdatedAt' => $value->externalData['passwordUpdatedAt'],
+                    ]
+                )
+            );
+
+            $this->repository->commit();
+        } catch (Exception $e) {
+            $this->repository->rollback();
+            throw $e;
+        }
+
+        return $this->loadUser($loadedUser->id);
```

Comment:
This is a lot of work for a simple password update. Is the only reason for it the validation? Can the validation of password be extracted to a separate validator and used in Content update and here?
This code should read:
1. Validate password
2. Update password

That's it.

Reply:
> Can the validation of password be extracted to a separate validator and used in Content update and here?

I agree.
Unfortunately I don't really think common code related to user update could be extracted and used in both - `ContentService` / `ContentValidator` do perform very specific checks and this version here already skips a few operations, mainly related to language of the content (which in this case isn't important).

Ideally I would only do checks against the new password, but at the point of writing I wasn't sure if password validation itself could be extracted / injected (circumventing the need to build the whole content update structure).

I'll look into it.

Reply:
@alongosz code has been revised in https://github.com/ezsystems/ezplatform-kernel/pull/117/commits/d2dec1a15187e202862d31b62d8341cb28b29df4. Now `updateUserPassword` does not rely on `ContentService` at all. I'm having doubts about how password hashing is done (it's a tad different from `UserPasswordEncoder` I'm more familiar with) so I'd ask you to see if I haven't made a mistake regarding selection of hash algorithm.